AudioSampleBufferCompressor and VideoSampleBufferCompressor are using sync dispatch

jyavenard opened the pull request:
#### f40822a656dc370b02a88bac25f1beb96ad9ad3f
<pre>
AudioSampleBufferCompressor and VideoSampleBufferCompressor are using sync dispatch
<a href="https://bugs.webkit.org/show_bug.cgi?id=281107">https://bugs.webkit.org/show_bug.cgi?id=281107</a>
<a href="https://rdar.apple.com/137560529">rdar://137560529</a>

Reviewed by Youenn Fablet.

Dispatch frame to be compressed to the encoding WorkQueue asynchronously.
To safely do so, make both AudioSampleBufferCompressor and VideoSampleBufferCompressor
inheriting ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::addSampleBuffer):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::create):
(WebCore::VideoSampleBufferCompressor::addSampleBuffer):

Canonical link: <a href="https://commits.webkit.org/284926@main">https://commits.webkit.org/284926@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/974ff197b6ea8526725eb7793afb6e9f270059d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70847 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74946 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22047 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72963 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58055 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21868 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56062 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14537 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61080 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36515 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42335 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18514 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20389 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76661 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18075 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63808 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15121 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61143 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63753 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11832 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5480 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/829 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47130 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48411 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46872 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -534,12 +534,9 @@
 {
-    // Heap allocations are forbidden on the audio thread for performance reasons so we need to
-    // explicitly allow the following allocation(s).
-    DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
```
We should probably keep these if we want to go back to WebProcess.

jyavenard replied:
My plan was to have the MediaRecorderPrivateWriter be in the main thread instead. so this would need to be in the thread proxy between the `RealtimeMediaSource::AudioSampleObserver` and `RealtimeMediaSource::VideoFrameObserver` and its MediaRecorderPrivate ; in which case it's there that this is needed.

jyavenard replied:
I've re-instated for now. This method is currently only ever called on the main thread

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -534,12 +534,9 @@
-    // explicitly allow the following allocation(s).
-    DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
-    m_serialDispatchQueue->dispatchSync([this, buffer] {
+    m_serialDispatchQueue->dispatch([this, buffer = RetainPtr { buffer }] {
```
There is no guarantee anymore that `this` stays alive given flush may not be called before its destructor.
I guess we could ensure that it stays valid by having a dispatchSync in its destructor.
The regular approach would probably be to make AudioSampleBufferCompressor thread safe ref counted.

jyavenard replied:
doh good point

jyavenard committed:
AudioSampleBufferCompressor and VideoSampleBufferCompressor are using sync dispatch
https://bugs.webkit.org/show_bug.cgi?id=281107
rdar://137560529

Reviewed by Youenn Fablet.

Dispatch frame to be compressed to the encoding WorkQueue asynchronously.
To safely do so, make both AudioSampleBufferCompressor and VideoSampleBufferCompressor
inheriting ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::addSampleBuffer):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::create):
(WebCore::VideoSampleBufferCompressor::addSampleBuffer):

Canonical link: https://commits.webkit.org/284926@main
